Added a Python library to faciliate creating the JSON-LD version of the test suite

gaurav opened the pull request:
The curation workflow now includes:
 * a Python library for modeling the phyloreference test suites (`add-labels/phyloref/*`), 
 * a paper-context.json file for converting it into RDF/XML (`./add-labels/paper-context.json`), 
 * and an OWL ontology that provides definitions for the terms we use (`./ontologies/phyloref_testcase.owl`)

We now use Taxonomic Units (TUs) for storing taxonomic information, compare TUs directly to identify matches, store matches in a Taxonomic Unit Match (TUMatch) object that includes information on why the match was made, and infer the relationship between specifer and node based on the TUMatches between their taxonomic units [using property chains](https://github.com/phyloref/curation-workflow/blob/develop/ontologies/phyloref_testcase.owl#L133).

This is now in line with [the blog post describing matching](https://github.com/phyloref/phyloref.github.io/pull/28), except for one detail:
 - [ ] The Python model currently treats scientific names as a separate entity, and a specimen to be a special type of taxonomic unit. In [the latest blog post draft](https://github.com/phyloref/phyloref.github.io/pull/28), I decided that it makes more sense to treat names/taxa as one type of taxonomic unit and specimens as the other kind. Once I make that change, this pull request will be ready for review!

We don't pass all tests for two reasons -- I will try to get these working this week, but we can merge this even if I'm unsuccessful:
 - [ ] [Some phyloreferences don't resolve](https://travis-ci.org/gaurav/curation-workflow/builds/320568249#L529) because they use specimen-based identifiers, which we don't support matching on yet.
 - [ ] [Others don't resolve](https://travis-ci.org/gaurav/curation-workflow/builds/320568249#L529) because some of the specifiers are missing in the provided phylogeny. We should probably modify jphyloref to mark those not as test failures but as untested.

hlapp commented:
There’s obviously way more here than can be effectively reviewed within s single pull request. So I’ll focus on a few technical things to move forward.

* Remove .DS_Store and add it to .gitignore. (And using `git add —all` is seldom a good idea.)
* add-labels seems to very misleading name (there’s no mention of “adding labels” in the description of the add-labels.py). Please change to an appropriate name, including the directory.
* The altering of sys.path at the beginning of add-labels.py ought to be unnecessary. Why is that needed, and can that be changed so that it’s not? It looks like quite a hack.

There are other things, such as jar files being duplicated and data files being stored that I assume are among the reasons this repo is starting to balloon in size, but we’ll probably best deal with those one at a time and outside of this mother of a PR.

gaurav commented:
Thanks, Hilmar! I'll fix those this week. I would definitely like to discuss the JAR file duplication and data file storage at some point, however!

gaurav commented:
Thanks, @hlapp! Here are the changes I've made:

 - Deleted .DS_Store and added it to .gitignore. (@9df49d4)
 - In fixing the name of "add-labels.py", I wanted to rename the term "test suite" I currently use to refer to the individual JSON files we process. I've renamed that to "test case", so now the entire curation workflow is a *test suite* that consists of individual *test cases* that contain *phylogeny groups* that contain  individual testable *phyloreferences*. Given those changes, I've renamed the script to `testcase2owl.py`. (@0277e24, @64b437e, @21b4a67)
 - You're right, the `sys.path` hack is no longer necessary! Deleted. (@8efc68d)

Please let me know what else I can improve!

hlapp commented:
My hunch is to suggest you squash merge this, but to keep salient changes in the commit message when revising / consolidating it. 

That's mainly because there's just so much here, and interspersed with merges from master that are distracting.

If you think there are changes here that you'd like to remain able to revert (or re-apply) by themselves rather than as the whole PR, then don't squash merge.

The best thing in terms of creating a useful history might be to rebase interactively the whole change set and consolidate into fewer individual commits by squashing sets of commits. But that's somewhat advanced work (`git rebase -i <initial-commit-on-this-changeset>`) and might give rise to rebase conflicts, so could get messy. Perhaps that adventure is better saved up for a less unwieldy PR.

gaurav commented:
@hlapp I've rebased this into a smaller set of changes and moved it to the phyloref organization. I'm going to open a new pull request for that branch (#10) and close this pull request, but please reopen it if you need to!
